File: EvaluateReversePolishNotation.py
class Solution:
    def evalRPN(self, tokens):
        """
        :type tokens: List[str]
        :rtype: int
        """
        stack = []
        for token in tokens:
            if token in ["+", "-", "*", "/"]:
                b, a = stack.pop(), stack.pop()
                if token == "+":
                    stack.append(a+b)
                elif token == "-":
                    stack.append(a-b)
                elif token == "*":
                    stack.append(a*b)
                elif token == "/":
                    stack.append(int(float(a)/b))  # stack.append(a//b) 得到的结果不同，6/(-132) 前者得到的结果是0, 而后者结果室-1
            else:
                stack.append(int(token))
        return stack[-1]
        
if __name__=='__main__':
    n = int(raw_input())
    tokens = []
    for i in range(n):
        s = raw_input()
        tokens.append(s)
    
    solution = Solution()
    print(solution.evalRPN(tokens))

File: EvaluateReversePolishNotation.cpp
#include<iostream>
#include<vector>
#include<string>

using namespace std;

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        
        vector<int> stack;
        for(int i=0;i<tokens.size();i++){
            if(tokens[i]=="+" || tokens[i]=="-" || tokens[i]=="*" || tokens[i]=="/"){
                int b = stack.back();
                stack.pop_back();
                int a = stack.back();
                stack.pop_back();
                int c;
                switch(tokens[i][0]){
                    case '+':
                        c = a + b;
                        break;
                    case '-':
                        c = a - b;
                        break;
                    case '*':
                        c = a * b;
                        break;
                    case '/':
                        c = a / b;
                        break;
                }
                stack.push_back(c);
            }else{
                stack.push_back(stringToInt(tokens[i]));
            }
        }
        
        return stack.back();
        
    }
    
    int stringToInt(string s){
        int num=0;
        int symble = 1;
        for(int i=0;i<s.length();i++){
            if(s[i]=='-'){
                symble = -1;
                continue;
            }
            num = num*10 + s[i]-'0';
        }
        
        return num*symble;
    }
};

int main(){
    int n;
    cin>>n;
    
    vector<string> tokens;
    for(int i=0;i<n;i++){
        string token;
        cin>>token;
        tokens.push_back(token);
    }
    
    Solution *solution = new Solution();
    cout<<solution->evalRPN(tokens);
    
    return 0;
}
